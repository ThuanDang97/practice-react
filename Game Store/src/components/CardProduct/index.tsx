import { Link } from 'react-router-dom'
import {
  Box,
  Card,
  CardContent,
  CardMedia,
  SvgIcon,
  Typography,
  useTheme,
} from '@mui/material'
import {
  RiGlobalLine,
  RiWindowsFill,
  RiAndroidFill,
  RiPlaystationFill,
  RiXboxFill,
  RiAppleFill,
  RiAddLine,
  RiCheckLine,
} from 'react-icons/ri'
import { SiIos, SiLinux, SiNintendoswitch } from 'react-icons/si'

// Components
import Button from '@components/Button'

// Hooks
import { useCartContext } from '@hooks/useCartContext'
import { useAuthContext } from '@hooks/useAuthContext'

// Types
import { Game } from '@self-types/Game.types'

export interface ICardProduct {
  game: Game
  addToCart: (game: Game) => void
}

const platformIcons: Record<string, React.ReactNode> = {
  web: <RiGlobalLine />,
  pc: <RiWindowsFill />,
  android: <RiAndroidFill />,
  ios: <SiIos />,
  playstation: <RiPlaystationFill />,
  xbox: <RiXboxFill />,
  mac: <RiAppleFill />,
  linux: <SiLinux />,
  nintendo: <SiNintendoswitch />,
}

const CardProduct = ({ game, addToCart }: ICardProduct) => {
  const theme = useTheme()
  const {
    id,
    name,
    price,
    released,
    background_image,
    parent_platforms,
    genres,
  } = game

  const releasedDate = new Date(released).toLocaleDateString()
  const genreList = genres.map((item) => item.name).join(', ')
  const { listCart } = useCartContext()
  const { userSession } = useAuthContext()

  const renderButton = () => {
    if (userSession) {
      return listCart.find((item) => item.id === id) ? (
        <Typography color="whitesmoke" fontWeight={700}>
          Added <RiCheckLine color="whitesmoke" />
        </Typography>
      ) : (
        <Button
          onClick={() => addToCart(game)}
          endIcon={<RiAddLine />}
          variant="text"
          hovercolor={theme.palette.secondary.main}
          sx={{
            padding: 0,
            fontWeight: 700,
            textTransform: 'none',
            color: 'grey',
          }}
        >
          Add to cart
        </Button>
      )
    }

    return (
      <Link to={'/login'}>
        <Typography color="whitesmoke" fontWeight={700} fontSize="15px">
          Please login now.
        </Typography>
      </Link>
    )
  }

  return (
    <Card sx={{ width: 300, borderRadius: 5, backgroundColor: '#1a1a1a' }}>
      <Link to={`/games/${game.id}`}>
        <CardMedia
          component="img"
          height="140"
          image={background_image}
          alt="green iguana"
        />
      </Link>
      <CardContent
        sx={{
          gap: 1.5,
          backgroundColor: '#1a1a1a',
          display: 'flex',
          flexDirection: 'column',
        }}
      >
        <CardContent
          sx={{ display: 'flex', justifyContent: 'space-between', padding: 0 }}
        >
          {renderButton()}
          <Typography color="gray">${price}</Typography>
        </CardContent>
        <Link to={`/games/${game.id}`}>
          <Typography
            variant="h4"
            fontSize={20}
            sx={{ color: theme.palette.common.white, fontWeight: 700 }}
          >
            {name}
          </Typography>
        </Link>
        <Box
          display="flex"
          flexDirection="column"
          gap={1.5}
          alignItems="flex-start"
        >
          {parent_platforms && (
            <Box display="flex" gap="10px">
              {parent_platforms.map(({ platform }) => (
                <SvgIcon
                  key={`${platform.id}-${id}`}
                  sx={{
                    color: theme.palette.common.white,
                    width: 16,
                    height: 16,
                  }}
                >
                  {platformIcons[platform.slug]}
                </SvgIcon>
              ))}
            </Box>
          )}
          <Typography variant="body2" color="gray">
            Released: {releasedDate}
          </Typography>
          <Typography variant="body2" color="gray">
            Genres: {genreList}
          </Typography>
        </Box>
      </CardContent>
    </Card>
  )
}

export default CardProduct
