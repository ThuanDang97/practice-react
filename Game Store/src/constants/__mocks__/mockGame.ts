export const MOCK_GAME_API = [
  {
    id: 3498,
    slug: 'grand-theft-auto-v',
    name: 'Grand Theft Auto V',
    ratings_count: 5927,
    description_raw: '',
    website: '',
    released: '2013-09-17',
    background_image:
      'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
    developers: [
      {
        name: '',
      },
    ],
    publishers: [
      {
        name: '',
      },
    ],
    parent_platforms: [
      {
        platform: {
          id: 186,
          name: 'Xbox Series S/X',
          slug: 'xbox-series-x',
        },
      },
    ],
    platforms: [
      {
        platform: {
          id: 1,
          name: 'PC',
          slug: 'pc',
        },
      },
    ],
    genres: [
      {
        name: 'Actions',
      },
    ],
    short_screenshots: [
      {
        id: -1,
        image:
          'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
      },
    ],
  },
]

export const MOCK_GAME_TRANSFORMED = [
  {
    id: 3498,
    slug: 'grand-theft-auto-v',
    name: 'Grand Theft Auto V',
    price: 19.99,
    ratings_count: 5927,
    description_raw: '',
    website: '',
    released: '2013-09-17',
    background_image:
      'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
    developers: [
      {
        name: '',
      },
    ],
    publishers: [
      {
        name: '',
      },
    ],
    parent_platforms: [
      {
        platform: {
          id: 186,
          name: 'Xbox Series S/X',
          slug: 'xbox-series-x',
        },
      },
    ],
    platforms: [
      {
        platform: {
          id: 1,
          name: 'PC',
          slug: 'pc',
        },
      },
    ],
    genres: [
      {
        name: 'Actions',
      },
    ],
    short_screenshots: [
      {
        id: -1,
        image:
          'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
      },
    ],
  },
]

export const MOCK_GAME_DATA = {
  id: 3498,
  slug: 'grand-theft-auto-v',
  name: 'Grand Theft Auto V',
  price: 19.99,
  ratings_count: 5927,
  description_raw: '',
  website: '',
  released: '2013-09-17',
  background_image:
    'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
  developers: [
    {
      name: '',
    },
  ],
  publishers: [
    {
      name: '',
    },
  ],
  parent_platforms: [
    {
      platform: {
        id: 186,
        name: 'Xbox Series S/X',
        slug: 'xbox-series-x',
      },
    },
  ],
  platforms: [
    {
      platform: {
        id: 1,
        name: 'PC',
        slug: 'pc',
      },
    },
  ],
  genres: [
    {
      name: 'Actions',
    },
  ],
  short_screenshots: [
    {
      id: -1,
      image:
        'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
    },
  ],
}
