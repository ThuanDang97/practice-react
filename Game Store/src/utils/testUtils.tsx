import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import { BrowserRouter as Router } from 'react-router-dom'

// Themes
import theme from '@self-types/Themes'

// Contexts
import { AuthProvider } from '@contexts/AuthProvider'
import { CartProvider } from '@contexts/CartProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <AuthProvider>
        <CartProvider>
          <Router>{component}</Router>
        </CartProvider>
      </AuthProvider>
    </ThemeProvider>,
  )
}

export * from '@testing-library/react'
export { customRender as render }
