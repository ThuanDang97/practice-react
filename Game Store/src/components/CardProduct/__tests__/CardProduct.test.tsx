// Mocks
import { MOCK_GAME_DATA } from '@constants/__mocks__/mockGame'

// Utils
import { render } from '@utils/testUtils'

// Components
import CardProduct, { ICardProduct } from '@components/CardProduct'

describe('Card Product Component', () => {
  const handleAddToCart = jest.fn()

  const props: ICardProduct = {
    game: MOCK_GAME_DATA,
    addToCart: handleAddToCart(),
  }

  it('should render Card Product', () => {
    const { container } = render(<CardProduct {...props} />)

    expect(container).toMatchSnapshot()
  })
})
