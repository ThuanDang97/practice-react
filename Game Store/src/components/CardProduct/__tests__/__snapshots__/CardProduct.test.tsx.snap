// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Product Component should render Card Product 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1rqr2o1-MuiPaper-root-MuiCard-root"
  >
    <a
      href="/games/3498"
    >
      <img
        alt="green iguana"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="140"
        src="https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg"
      />
    </a>
    <div
      class="MuiCardContent-root css-1n6x7yt-MuiCardContent-root"
    >
      <div
        class="MuiCardContent-root css-1xpvmq8-MuiCardContent-root"
      >
        <a
          href="/login"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-t0gjje-MuiTypography-root"
          >
            Please login now.
          </p>
        </a>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1rfn1w6-MuiTypography-root"
        >
          $
          19.99
        </p>
      </div>
      <a
        href="/games/3498"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-1dl0pq9-MuiTypography-root"
        >
          Grand Theft Auto V
        </h4>
      </a>
      <div
        class="MuiBox-root css-19n1w9z"
      >
        <div
          class="MuiBox-root css-i5q2k0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-216ivs-MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          />
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-98hkvb-MuiTypography-root"
        >
          Released: 
          9/17/2013
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 css-98hkvb-MuiTypography-root"
        >
          Genres: 
          Actions
        </p>
      </div>
    </div>
  </div>
</div>
`;
